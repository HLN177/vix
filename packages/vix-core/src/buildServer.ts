import builtinModules from 'builtin-modules';
import colors from 'picocolors';
import { analyzeMetafile } from 'esbuild';
import path from 'path';
import fse from 'fs-extra';
import deepmerge from 'deepmerge';
import { build, BuildOptions } from 'esbuild';
import resovleAlias from 'esbuild-plugin-path-alias';
import { ifDependenceExist } from './utils';
import {
  ServerBuildOption,
  FastUserConfig,
} from './plugins/browserBuildPlugin/types';
import { Logger } from 'vite';

const graphqlPlugin = (gql) =>
  gql &&
  require('@luckycatfactory/esbuild-graphql-loader').default({
    filterRegex: /\.(gql|graphql)$/,
  });

export function cleanOptions<Options extends ServerBuildOption>(
  options: Options
): Omit<Options, keyof BuildOptions> {
  delete options.entry;
  delete options.resolve;
  delete options.plugins;
  return options;
}

const buildBundle = async (options: FastUserConfig) => {
  const { serverBuild, presets = {} } = options;
  const {
    entry,
    outdir,
    resolve = {},
    plugins: customPlugins,
  } = serverBuild || {};
  const { plugins: presetsPlugins = {} } = presets;
  const { graphql = ifDependenceExist('graphql') } = presetsPlugins;
  const { alias } = resolve;
  const entryPoints = Array.isArray(entry) ? entry : [entry];
  const config: any = deepmerge(
    {
      platform: 'node',
      allowOverwrite: true,
      entryPoints,
      outdir,
      minify: true,
      metafile: true,
      absWorkingDir: process.cwd(),
      target: ['node12'],
      external: builtinModules,
      format: 'cjs',
      bundle: true,
      plugins: [alias && resovleAlias(alias), graphqlPlugin(graphql)]
        .concat(customPlugins)
        .filter(Boolean),
    },
    cleanOptions(serverBuild)
  );
  return build(config);
};


export default async (config: FastUserConfig, logger: Logger) => {
  let report;
  const { serverBuild } = config;
  const { metafile, entry, ...restServerBuild } = serverBuild!;
  let serverBuildOptions: any = restServerBuild;
  let metaOutputFile: string | null = null;
  let outdir: string | null = null;

  if (Array.isArray(entry)) {
    outdir = serverBuild!.outdir || './dist';
    metaOutputFile = 'build.meta.json';
    serverBuildOptions = { ...restServerBuild, outdir, entry };
  } else {
    const { outfile } = serverBuild!;
    const fileName: string = (outfile || (entry as string))!;
    const distFileName = path.parse(fileName).name;
    metaOutputFile = `${distFileName}.meta.json`;
    const outputFile = outfile || `./dist/${distFileName}.js`;
    outdir = path.dirname(outputFile);
    serverBuildOptions = {
      ...restServerBuild,
      outfile: outputFile,
      entry,
    };
  }

  const result = await buildBundle({
    ...config,
    serverBuild: serverBuildOptions,
  });
  if (metafile) {
    report = await analyzeMetafile(result.metafile!, {
      color: true,
    });
    logger.info(report);
    await fse.writeJSON(outdir + '/' + metaOutputFile, result.metafile);
  }
  const printReport = (metafile: any) => {
    logger.info(colors.bold('Server Build:'));
    const { outputs } = metafile;
    const indents = '               ';
    const formatOutput = (key: string) =>
      logger.info(
        colors.green(
          `${key}${indents}${outputs[key].bytes / 1000}Kib`
        )
      );
    for (const key of Object.keys(outputs)) {
      formatOutput(key);
    }
    logger.info('');
  };
  result?.metafile && printReport(result.metafile);
}