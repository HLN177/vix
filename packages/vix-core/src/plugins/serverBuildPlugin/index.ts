import type { Logger, Plugin } from 'vite';
import buildServer from '../../buildServer';
import { CLI_ALIAS } from '../../types';
import { FastUserConfig } from '../browserBuildPlugin/types'

export const PLUGIN_NAME = `${CLI_ALIAS}:server-build-plugin`;

function serverBuildPlugin(cfg: FastUserConfig = {}): Plugin {
    const fastConfig: FastUserConfig = cfg;
    let logger: Logger;
    let mode: string;
    return <Plugin>{
        name: PLUGIN_NAME,
        enforce: 'post',
        apply: 'build',
        config: (config) => {
            const { customLogger, mode: optionMode } = config;
            logger = customLogger;
            mode = optionMode;
        },
        closeBundle: async () => {
            const { serverBuild } = fastConfig;
            if ([serverBuild, serverBuild?.entry].every(Boolean)) {
                await buildServer(fastConfig, logger, mode);
            }
        }
    };
}

export default serverBuildPlugin;
