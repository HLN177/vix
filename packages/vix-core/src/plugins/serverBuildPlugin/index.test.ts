import { vi, describe, expect, test, beforeEach, afterEach } from 'vitest';
import serverBuildPlugin, {PLUGIN_NAME} from './index';
import buildServer from '../../buildServer';
import type { Plugin } from 'vite';

describe('Return As Plugin', () => {
  let plugin: Plugin;

  beforeEach(() => {
    plugin = serverBuildPlugin({});
  });

  test('Should return Object Structure keys', () => {
    ['name', 'enforce', 'apply', 'config', 'closeBundle'].forEach(key => expect(plugin.hasOwnProperty(key)).toBeTruthy());
  });

  test('Should check name, enforce and apply are string type', () => {
    ['name', 'enforce', 'apply'].forEach(key => expect(plugin[key]).toBeTypeOf('string'));
  });

  test('Should check config, and closeBunlde are function type', () => {
    ['config', 'closeBundle'].forEach(key => expect(plugin[key]).toBeTypeOf('function'));
  });
});

describe('serverBuildPlugin Function()', () => {
  test('Should return an object with expected value', () => {
    const plugin = serverBuildPlugin({});
    expect(plugin.name).toBe(PLUGIN_NAME);;
    expect(plugin.enforce).toBe('post');;
    expect(plugin.apply).toBe('build');;
  });

  test('Should return an object matches snapshot', () => {
    const plugin = serverBuildPlugin({});
    expect(plugin).toMatchSnapshot();
  });
});

vi.mock('../../buildServer', () => {
  return {
    default: vi.fn()
  };
});

describe('Test serverBuildPlugin().config() and serverBuildPlugin().closeBundle()', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  const mockOption = {
    customLogger: 'fakelogger',
    mode: 'fakeMode'
  };

  describe('Test fastconfig.serverBuild value', () => {
    test('When its entry property has truthy value, should call buildServer()', async () => {
      const mockConfig = {
        serverBuild: {
          entry: 'mockEntry'
        }
      };
      const plugin1: any = serverBuildPlugin(mockConfig);
      plugin1.config(mockOption);
      await plugin1.closeBundle();
      expect(buildServer).toBeCalledWith(mockConfig, mockOption.customLogger, mockOption.mode) 
    });

    test('When its entry property has falsy value, should not call buildServer()', async () => {
      const mockConfig = {
        serverBuild: {
          entry: ''
        }
      };
      const plugin1: any = serverBuildPlugin(mockConfig);
      plugin1.config(mockOption);
      await plugin1.closeBundle();
      expect(buildServer).not.toHaveBeenCalled();
    });

    test('When it doesn\'t have entry property, should not call buildServer()', async () => {
      const mockConfig = {
        serverBuild: ''
      };
      const plugin1: any = serverBuildPlugin(mockConfig as any);
      plugin1.config(mockOption);
      await plugin1.closeBundle();
      expect(buildServer).not.toHaveBeenCalled();
    });
  });
});