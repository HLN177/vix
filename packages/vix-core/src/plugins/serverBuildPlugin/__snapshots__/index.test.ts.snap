// Vitest Snapshot v1

exports[`serverBuildPlugin Function() > Should return an object matches snapshot 1`] = `
{
  "apply": "build",
  "closeBundle": [Function],
  "config": [Function],
  "enforce": "post",
  "name": "vix:server-build-plugin",
}
`;
